Finishes carnmc's shuttle rotation feature.

MrsTonedOne opened the pull request:
Under carn shuttles, When shuttles dock at a docking port that is not the same direction as them, they rotate their bounding box, and they do magic i don't understand to rotate where the turfs get placed.

But it doesn't rotate the items or smoothwall or handle pixel_xy or any of that.

This fixes that.

Shuttle rotation will call /atom/proc/shuttleRotate(degrees)
Default behavior handles most cases, and overrides handle edge cases.

Shuttle walls don't smooth wall or obey directions, so I just rotate their icon using matrix transform, its dirty but it works.

![space station 13 2015-07-19 001851](https://cloud.githubusercontent.com/assets/7069733/8765294/86fca0da-2dba-11e5-9ad0-dd5ef5287c0f.png)
![space station 13 2015-07-19 013657](https://cloud.githubusercontent.com/assets/7069733/8765295/870032b8-2dba-11e5-838f-c5e1c0937248.png)

To utilize, just change one of the shuttle's stationary docking port's dir to anything besides the shuttle's mobile docking port's dir

On box, the station docking port for the eshuttle is in the space turf just right of the top general access airlock (the one below the sec airlock)

On meta, the station docking port for the eshuttle is in the space turf just below the right most airlock

View the turf's contents in VV to see it.

To make the second picture, I moved the station's docking port object to inside the cargo external airlock's turf (not inside the airlock itself, that would break shit) then I changed its dir from east to north.

MrsTonedOne committed:
Finishes carnmc's shuttle rotation feature.

Under carn shuttles, When shuttles dock at a docking port that is not the same direction as them, they rotate their bounding box, and they do magic i don't understand to rotate where the turfs get placed.

But it doesn't rotate the items or smoothwall or handle pixel_xy or any of that.

This fixes that.

Shuttle rotation will call /atom/proc/shuttleRotate(degrees)
Default behavior handles most cases, and overrides handle edge cases.

Shuttle walls don't smooth wall or obey directions, so I just rotate their icon using matrix transform, its dirty but it works.

See: https://tgstation13.org/msoshit/powrightinthekisser.png or https://tgstation13.org/msoshit/fuckingshuttles.png

RemieRichards commented on code/__HELPERS/game.dm:
```diff
@@ -379,3 +379,39 @@
+			return WEST
+
+
+/proc/getDegreesFromDir(dir)
```
dir2angle() exists, returns the same values as yours, and includes diags :P

MrStonedOne replied:
including diags is actually unwanted for this, would cause issues.

RemieRichards replied:
you're never inputting a diagonal dir nerd.

MrStonedOne replied:
it could happen, but anywho, I did what you asked and have it set to round up if the rotation angle isn't a mod of 90.

Dorsisdwarf commented:
I like this, so we can have the shuttles dock more fluidly / inflight reorientation to disorientate people

Jalleo commented:
This means we can merge the cargo shuttle on all the maps one variant one size for all.

MrsTonedOne committed:
Shuttle Rotation Tweaks

This should fix shuttle walls not rotating back to their old state correctly when the shuttle docks elsewhere.
Also uses built in procs and caches the calculated rotation angle to improve shuttle docking speed by not calculating it again for every object.
This should also ensure that diagonal objects rotate correctly, rather then snapping back to a cardinal direction.

MrStonedOne commented:
Jalleo one of my next changes to shuttles is gonna be to have all the centcom shuttles start on the station zlevel and get moved during pre-game, so we can stop modifying the centcom z level in every map

Metacide commented:
Stellar, so this sort of issue won't happen again?

![did not expect](https://cloud.githubusercontent.com/assets/3833444/8751348/4fbbb2a8-2ca4-11e5-9718-b6fec45f0088.png)

MrStonedOne commented:
yep! I already tested meta shuttle as well, it works 100%

in fact, let me get you some pics of that

Fayrik commented:
I don't want to overhype this, but this might just be feature of the year, here.

I'm guessing it's a map issue, but the thrusters in those screenshots are facing the wrong way, also.

Even if it is a bug, it's still amazing enough to beat any other SS13 shuttle system ever. I love it.

Aranclanos commented:
>and they do magic i don't understand to rotate where the turfs get placed

#worried

Fayrik commented:
@Aranclanos @MrStonedOne Isn't that "Magic" just simple trigonometry? I'm pretty sure that's what the custom sin and cos functions exist for.

MrStonedOne commented:
>I'm guessing it's a map issue, but the thrusters in those screenshots are facing the wrong way, also.

Sprite issue, the directional sprites are off easy fix, but not my job.

Xhuis commented:
![kreyg](https://cloud.githubusercontent.com/assets/5659269/8771597/8f4e91b2-2e91-11e5-9d7f-eadb15e4cbd4.png)

xxalpha commented on code/modules/shuttle/shuttle.dm:
```diff
@@ -324,54 +360,49 @@
+				if(istype(M, /mob/living/carbon))
+					if(!M.buckled)
+						M.Weaken(3)
 
```
Somewhere around here add this loop:

```
for(var/turf/T in L1)
		for(var/atom/A in T)
			if(A.smooth)
				smooth_icon(A)
```

This will resmooth all smooth atoms *after* they're moved. Smoothing onboard the shuttle was working before because of the spawn in the smooth code, which was removed. This is needed now.

MrStonedOne replied:
Alright, I'll run a rebase and test this out..

how ever, shuttle code doesn't need to resmooth when it moves, only when it rotates, so i'll just add this to atom/shuttleRotate

Jalleo replied:
Probanly because byond thinks its making a new wall instead of moving one?

MrStonedOne replied:
This happens to tables as well.. even when no rotation happens.

![image](https://cloud.githubusercontent.com/assets/7069733/8776902/42e3c0e6-2ea9-11e5-83ae-2455a1fa4bc1.png)

optimumtact commented:
Why not just resmooth when the shuttle moves?

MrsTonedOne committed:
Makes shuttle rotate Re-Smooth's atoms if need be.

MrsTonedOne committed:
fixes shuttle rotate compile error
